services/dynamic-scheduler/tests/unit/service_tracker/test__api.py

```diff
@@ -0,0 +1,321 @@
+from models_library.services_enums import ServiceState
+from pydantic import NonNegativeInt
+from pytest_simcore.helpers.typing_env import EnvVarsDict
+from servicelib.deferred_tasks import TaskUID
+from servicelib.utils import logged_gather
+from settings_library.redis import RedisSettings
+from simcore_service_dynamic_scheduler.services.service_tracker import (
+    get_all_tracked,
+    get_tracked,
+    remove_tracked,
+    set_if_status_changed,
+    set_request_as_running,
+    set_request_as_stopped,
+    set_service_status_task_uid,
+)
+from simcore_service_dynamic_scheduler.services.service_tracker._api import (
+    _LOW_RATE_POLL_INTERVAL,
+    NORMAL_RATE_POLL_INTERVAL,
+    _get_current_state,
+    _get_poll_interval,
+)
+from simcore_service_dynamic_scheduler.services.service_tracker._models import (
+    SchedulerServiceState,
+    UserRequestedState,
+)
+
+pytest_simcore_core_services_selection = [
+    "redis",
+]
+
+
+@pytest.fixture
+def app_environment(
+    disable_rabbitmq_setup: None,
+    disable_deferred_manager_setup: None,
+    disable_notifier_setup: None,
+    app_environment: EnvVarsDict,
+    redis_service: RedisSettings,
+    remove_redis_data: None,
+) -> EnvVarsDict:
+    return app_environment
+
+
+@pytest.fixture
+def node_id() -> NodeID:
+    return uuid4()
+
+
+async def test_services_tracer_set_as_running_set_as_stopped(
+    app: FastAPI,
+    node_id: NodeID,
+    get_dynamic_service_start: Callable[[NodeID], DynamicServiceStart],
+    get_dynamic_service_stop: Callable[[NodeID], DynamicServiceStop],
+):
+    async def _remove_service() -> None:
+        await remove_tracked(app, node_id)
+        assert await get_tracked(app, node_id) is None
+        assert await get_all_tracked(app) == {}
+
+    async def _set_as_running() -> None:
+        await set_request_as_running(app, get_dynamic_service_start(node_id))
+        tracked_model = await get_tracked(app, node_id)
+        assert tracked_model
+        assert tracked_model.requested_state == UserRequestedState.RUNNING
+
+    async def _set_as_stopped() -> None:
+        await set_request_as_stopped(app, get_dynamic_service_stop(node_id))
+        tracked_model = await get_tracked(app, node_id)
+        assert tracked_model
+        assert tracked_model.requested_state == UserRequestedState.STOPPED
+
+    # request as running then as stopped
+    await _remove_service()
+    await _set_as_running()
+    await _set_as_stopped()
+
+    # request as stopped then as running
+    await _remove_service()
+    await _set_as_stopped()
+    await _set_as_running()
+
+
+@pytest.mark.parametrize("item_count", [100])
+async def test_services_tracer_workflow(
+    app: FastAPI,
+    node_id: NodeID,
+    item_count: NonNegativeInt,
+    get_dynamic_service_start: Callable[[NodeID], DynamicServiceStart],
+    get_dynamic_service_stop: Callable[[NodeID], DynamicServiceStop],
+):
+    # ensure more than one service can be tracked
+    await logged_gather(
```

Comment:
I did not check everywhere in your code, but limited_gather is better if you expect really a load of calls (>100)